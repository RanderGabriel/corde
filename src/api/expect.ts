import { expectMatchWithNot, MatchesWithNot } from "./matcher";

/**
 * Receives wich command will be tested.
 *
 * @param commandName Command name.
 *
 * @description Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @returns The **Compare** object, where will handle
 * the type of response is expected.
 */
export function expect(commandName: string): MatchesWithNot {
  return expectMatchWithNot(commandName);
}
