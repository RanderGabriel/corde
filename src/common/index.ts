export * from './cordeBot';
export * from './runtime';
export * from './testColletor';
