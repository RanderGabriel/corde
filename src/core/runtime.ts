import * as Discord from 'discord.js';
import ConfigOptions from '../models';
import CordeBot from './cordeBot';

export const DEFAULT_TEST_TIMEOUT = 5000;

/**
 * Default interface of json config
 *
 * @description `botTestToken` is not required.
 * only inform if is desired to start test bot with corde
 *
 * @implements ConfigOptions
 */
class Config implements ConfigOptions {
  cordeTestToken: string;
  botTestId: string;
  botTestToken?: string;
  channelId: string;
  guildId: string;
  timeOut?: number;
  botPrefix: string;
  testFilesDir: string;
  message: Discord.Message;
  files: string[];
}

class Runtime {
  configs: Config;
  bot: CordeBot;

  constructor() {
    this.bot = new CordeBot();
    this.configs = new Config();
  }

  setConfigs(configs: ConfigOptions) {
    this.configs.botPrefix = configs.botPrefix;
    this.configs.botTestId = configs.botTestId;
    this.configs.botTestToken = configs.botTestToken;
    this.configs.channelId = configs.channelId;
    this.configs.cordeTestToken = configs.cordeTestToken;
    this.configs.guildId = configs.guildId;
    this.configs.testFilesDir = configs.testFilesDir;
    this.configs.timeOut = configs.timeOut;
    this.configs.files = [];
  }

  loadBotSettings() {
    this.bot.loadChannel(this.guildId, this.channelId);
  }

  get channelId() {
    return this.configs.channelId;
  }

  get guildId() {
    return this.configs.guildId;
  }

  get obverseBotStart() {
    return this.bot.hasInited;
  }
}

const runtime = new Runtime();
export default runtime;
