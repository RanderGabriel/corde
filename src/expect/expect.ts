import { VoidLikeFunction } from "../types";
import { ExpectMatchesWithNot, MatchWithNot } from "./matcher";

/**
 * Invalid declaration of a test matcher.
 *
 * @since 3.0
 */
export function expect(): void;
/**
 * Receives which command will be tested.
 *
 * Do not inform the command prefix if
 * it's already informed in **configs**
 *
 * @param commandNameResolvable Command name. (Empty strings will resolve failed test)
 *
 * @returns An object with all possible tests to be done
 * in the bot.
 *
 * @since 1.0
 */
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable: T,
): MatchWithNot;
export function expect<T extends VoidLikeFunction | number | string>(
  commandNameResolvable?: T,
): MatchWithNot | void {
  if (commandNameResolvable == null) {
    return;
  }
  return new ExpectMatchesWithNot(commandNameResolvable);
}
