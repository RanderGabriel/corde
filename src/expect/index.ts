import { RoleMatches, MessageMatches } from "./matcher";
import { Expect, MacherContructorArgs } from "./types";

function getMessageMatchers(): string[] {
  return getFunctions(MessageMatches);
}

function getRoleMatchers(): string[] {
  return getFunctions(RoleMatches);
}

function getFunctions<T>(type: new (...args: any[]) => T): string[] {
  return Object.getOwnPropertyNames(type.prototype).filter(
    (propName) => propName !== "constructor" && typeof type.prototype[propName] === "function",
  );
}

const expectation = {
  not: {},
  inChannel: {},
  inGuild: {},
};

function createTestFunction(classType: any, params: MacherContructorArgs, functionName: string) {
  return (...args: any[]) => (new classType(params) as any)[functionName](...args);
}

function createTestsFromMatches<T>(
  names: string[],
  params: MacherContructorArgs,
  classType: new (...args: any[]) => T,
): [string, (...args: any[]) => any][] {
  return names.map((name) => {
    return [name, createTestFunction(classType, params, name)];
  });
}

function set(from: [string, (...args: any[]) => any][], to: any) {
  from.forEach((test) => {
    to[test[0]] = test[1];
  });
}

const messageTestNames = getMessageMatchers();
const roleMatchers = getRoleMatchers();

const _expect: any = <T extends (() => number | string) | number | string>(commandName: T) => {
  const messageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: false },
    MessageMatches,
  );

  const isNotMessageTests = createTestsFromMatches(
    messageTestNames,
    { commandName, isNot: true },
    MessageMatches,
  );

  const roleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: false },
    RoleMatches,
  );

  const isNotRoleTests = createTestsFromMatches(
    roleMatchers,
    { commandName, isNot: true },
    RoleMatches,
  );

  set(messageTests, expectation);
  set(isNotMessageTests, expectation.not);

  set(roleTests, expectation);
  set(isNotRoleTests, expectation.not);

  expectation.inChannel = (channelId: string) => {
    const channelMatchers: any = {
      not: {},
    };

    const messageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId },
      MessageMatches,
    );

    const isNotmessageTests = createTestsFromMatches(
      messageTestNames,
      { commandName, channelId, isNot: true },
      MessageMatches,
    );

    set(messageTests, channelMatchers);
    set(isNotmessageTests, channelMatchers.not);

    return channelMatchers;
  };

  expectation.inGuild = (guildId: string) => {
    const guildMatchers: any = {
      not: {},
    };

    const roleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    const isNotRoleTests = createTestsFromMatches(
      messageTestNames,
      { commandName, guildId },
      RoleMatches,
    );

    set(roleTests, guildMatchers);
    set(isNotRoleTests, guildMatchers.not);

    return guildMatchers;
  };

  return expectation;
};

const messageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: false, isCascade: true },
  MessageMatches,
);

const isNotMessageTests = createTestsFromMatches(
  messageTestNames,
  { commandName: "", isNot: true, isCascade: true },
  MessageMatches,
);

const roleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: false, isCascade: true },
  RoleMatches,
);

const isNotRoleTests = createTestsFromMatches(
  roleMatchers,
  { commandName: "", isNot: true, isCascade: true },
  RoleMatches,
);

_expect.not = {};

set(messageTests, _expect);
set(isNotMessageTests, _expect.not);

set(roleTests, _expect);
set(isNotRoleTests, _expect.not);

const expect = _expect as Expect;

export { expect };
