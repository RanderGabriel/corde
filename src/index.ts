/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
