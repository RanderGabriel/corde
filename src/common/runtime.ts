import { CordeBotClient } from "../core";
import ConfigOptions, { CordeBot } from "../interfaces";
import { Config } from "./config";

class Runtime {
  public bot: CordeBot;
  public configFilePath: string;
  public files: string[];
  public configs: Config;

  constructor() {
    this.configs = new Config();
  }

  public setConfigs(configs: ConfigOptions) {
    if (!configs) {
      throw new Error("Invalid Configs");
    }

    if (!this.configs) {
      this.configs = new Config();
    }
    this.configs = configs;
    this.loadBot();
  }

  private loadBot() {
    this.bot = new CordeBotClient(
      this.configs.botPrefix,
      this.configs.guildId,
      this.configs.channelId,
      this.configs.timeOut,
      this.configs.botTestId,
    );
  }
}

const runtime = new Runtime();
export { runtime };
