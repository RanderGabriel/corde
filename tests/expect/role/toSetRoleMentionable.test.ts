import { Client } from "discord.js";
import { ToSetRoleMentionable } from "../../../src/expect/matches";
import MockDiscord from "../../mocks/mockDiscord";
import { initCordeClientWithChannel } from "../../testHelper";
import { TestReport } from "../../../src/types";
import { buildReportMessage } from "../../../src/utils";
import { MockEvents } from "../../mocks/mockEvents";
import { runtime } from "../../../src/common/runtime";

let mockDiscord = new MockDiscord();

function initClient() {
  const corde = initCordeClientWithChannel(mockDiscord, new Client());
  corde.findRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.fetchRole = jest.fn().mockReturnValue(mockDiscord.role);
  corde.sendTextMessage = jest.fn().mockImplementation(() => {});
  return corde;
}

describe("testing toSetRoleMentionable operation", () => {
  afterEach(() => {
    mockDiscord = new MockDiscord();
  });

  it("should fail due to undefined roleIdentifier", async () => {
    const corde = initClient();
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    const report = await toSetRoleMentionable.action(true, undefined);

    const message = buildReportMessage(
      "expected: data to identifier the role (id or name)\n",
      `received: null`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid mentionable parameter (object)", async () => {
    const corde = initClient();
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    // @ts-ignore
    const report = await toSetRoleMentionable.action({}, { id: "123" });

    const message = buildReportMessage(
      `expected: mentionable parameter to be of boolean type\n`,
      `received: object`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid mentionable parameter (undefined)", async () => {
    const corde = initClient();
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    // @ts-ignore
    const report = await toSetRoleMentionable.action(undefined, { id: "123" });

    const message = buildReportMessage(
      `expected: mentionable option to be true or false\n`,
      `received: undefined`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to not found role", async () => {
    const corde = initClient();
    corde.findRole = jest.fn().mockReturnValue(null);
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    const report = await toSetRoleMentionable.action(false, { id: "123" });

    const message = buildReportMessage(`expected: role with id 123\n`, `received: null`);

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to no role mentionable was not updated", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);

    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    const report = await toSetRoleMentionable.action(false, { id: "123" });

    const message = buildReportMessage(
      `expected: mentionable to be false\n`,
      `received: role mentionable was not updated`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to isNot true and no role change", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);

    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", true);
    const report = await toSetRoleMentionable.action(false, { id: "123" });

    const expectReport: TestReport = {
      pass: true,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to role changed the mentionable (isNot false)", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    mockEvent.mockOnceMentionableUpdate(mockDiscord.role);
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    const report = await toSetRoleMentionable.action(mockDiscord.role.mentionable, { id: "123" });

    const expectReport: TestReport = {
      pass: true,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to mentionable should not change (isNot true)", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    mockEvent.mockOnceMentionableUpdate(mockDiscord.role);
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", true);
    const report = await toSetRoleMentionable.action(mockDiscord.role.mentionable, { id: "123" });

    const message = buildReportMessage(
      `expected: mentionable to not be ${mockDiscord.role.mentionable}\n`,
      `received: ${mockDiscord.role.mentionable}`,
    );

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due expected name did not match to received", async () => {
    const corde = initClient();

    runtime.setConfigs({ timeOut: 100 }, true);
    const mockEvent = new MockEvents(corde, mockDiscord);
    mockEvent.mockOnceMentionableUpdate(mockDiscord.role);
    const toSetRoleMentionable = new ToSetRoleMentionable(corde, "test", false);
    const report = await toSetRoleMentionable.action(true, { id: "123" });

    const message = buildReportMessage(`expected: mentionable to be true\n`, `received: false`);

    const expectReport: TestReport = {
      pass: false,
      message,
    };

    expect(report).toEqual(expectReport);
    expect(report).toMatchSnapshot();
  });
});
