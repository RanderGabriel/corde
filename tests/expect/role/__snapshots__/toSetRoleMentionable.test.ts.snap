// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toSetRoleMentionable operation should fail due to no role mentionable was not updated 1`] = `
Object {
  "message": "    
    expected: mentionable to be false
    received: role mentionable was not updated
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should fail due to undefined roleData 1`] = `
Object {
  "message": "    
    expected: data to identifier the role (id or name)
    received: null
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should return a not passed test due expected name did not match to received 1`] = `
Object {
  "message": "    
    expected: mentionable to be true
    received: false
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should return a not passed test due to mentionable should not change (isNot true) 1`] = `
Object {
  "message": "    
    expected: mentionable to not be false
    received: false
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should return false due to invalid mentionable parameter (object) 1`] = `
Object {
  "message": "    
    expect: mentionable parameter to be of boolean type
    received: object
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should return false due to invalid mentionable parameter (undefined) 1`] = `
Object {
  "message": "    
    expected: mentionable option to be true or false
    received: undefined
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should return false due to not found role 1`] = `
Object {
  "message": "    
    expected: role with id 123
    received: null
",
  "pass": false,
}
`;

exports[`testing toSetRoleMentionable operation should return true due to isNot true and no role change 1`] = `
Object {
  "pass": true,
}
`;

exports[`testing toSetRoleMentionable operation should return true due to role changed the mentionable (isNot false) 1`] = `
Object {
  "pass": true,
}
`;
